File: src/agent/src/Safir.Agent/Configuration.fs
namespace Safir.Agent.Configuration

open Microsoft.Extensions.Configuration
open Microsoft.Extensions.Options
open Safir.Messaging.Configuration

type AgentOptions() =
    member val DataDirectory: string = null with get, set
    member val EnableGrpcReflection = false with get, set
    member val EnableSwagger = false with get, set
    member val MaxDepth = 0 with get, set
    member val Redis = "" with get, set

type SafirMessaging(configuration: IConfiguration) =
    interface IConfigureOptions<MessagingOptions> with
        member this.Configure(options) =
            options.ConnectionString <- configuration.Get<AgentOptions>().Redis

File: src/agent/src/Safir.Agent/Configuration.cs
using Microsoft.Extensions.Options;

namespace Safir.Agent;

internal sealed class AgentConfiguration
{
    public string? DataDirectory { get; set; }
}

internal sealed record AgentOptions
{
    public required string DataDirectory { get; init; }
}

internal sealed record FileWatcherOptions
{
    public required string Path { get; set; }
}

internal static class OptionsExtensions
{
    public static AgentOptions Parse(this AgentConfiguration? configuration)
    {
        ArgumentNullException.ThrowIfNull(configuration);
        ArgumentException.ThrowIfNullOrEmpty(configuration.DataDirectory);

        return new AgentOptions {
            DataDirectory = configuration.DataDirectory,
        };
    }

    public static AgentOptions Parse(this IOptions<AgentConfiguration> options)
        => Parse(options.Value);
}
